File: 278-first-bad-version/278-first-bad-version.java
/* The isBadVersion API is defined in the parent class VersionControl.
      boolean isBadVersion(int version); */

public class Solution extends VersionControl 
{
    public int firstBadVersion(int n) 
    {
        // Accounts for the case when bad = 1
        if(isBadVersion(1))
            return 1;
        
        int low = 2;
        int high = n;
        
        while(low <= high)
        {
            int mid = low + (high - low) / 2;
            boolean isMidBad = isBadVersion(mid);
            
            if(isMidBad && !isBadVersion(mid - 1))
                return mid;
            
            if(isMidBad)
                high = mid - 1;
            else
                low = mid + 1;
        }
        
        return -1;
    }
}

File: 278-first-bad-version/278-first-bad-version.cpp
// The API isBadVersion is defined for you.
// bool isBadVersion(int version);

class Solution 
{
public:
    int firstBadVersion(int n) 
    {
        int low = 1;
        int high = n;
        int mid;
        while(low <= high)
        {
            mid = low + (high - low) / 2;
            bool isBad = isBadVersion(mid);
            if(isBad && !isBadVersion(mid - 1))
                return mid;
            if(isBad)
                high = mid - 1;
            else
                low = mid + 1;
        }
        
        return -1;
    }
};